plugins/metrics/pkg/agent/runner.go

```diff
@@ -148,6 +169,32 @@ func (tr *taskRunner) OnTaskPending(_ context.Context, _ task.ActiveTask) error
 	return nil
 }
 
+func (tr *taskRunner) doPush(ctx context.Context, writeRequest *prompb.WriteRequest) error {
+	expbackoff := tr.backoffPolicy.Start(ctx)
+
+	for {
+		select {
+		case <-expbackoff.Done():
+			return fmt.Errorf("failed to push data after exponential backoff")
+		case <-expbackoff.Next():
+			var err error
+
+			tr.remoteWriteClient.Use(func(remoteWriteClient remotewrite.RemoteWriteClient) {
+				promClient := remotewrite.AsPrometheusRemoteWriteClient(remoteWriteClient)
+				_, err = promClient.Push(ctx, writeRequest)
+			})
+
+			if !isRecoverable(err) {
```

Comment:
Just wondering what the behaviour is here when the gateway/cortex restarts? is it context cancelled

Reply:
If ctx is canceled, the backoff policy will be canceled and the task will go into the failed state.

Reply:
sorry, to clarify, this appears to be incorrect logic.

Reply:
Shoot you're right. We need to make the distinction between the push context failing and the task context failing. I'll push a fix in a bit.